### PR TITLE
[Pubsub] Generalize pubsub impl part 1

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is reviewable. Currently failures are all cpp unit test failures (which is expected)

Doing 2 things.

1. Use UniqueID for pub/sub IDs.
2. change the failure handling mechanism to timeout based.

## Related issue number

https://github.com/ray-project/ray/issues/14322

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/pubsub/publisher.h
Comment (rkooo567) [inline src/ray/core_worker/pubsub/publisher.h:22]: Need to remove it.

File: src/ray/core_worker/pubsub/publisher.h
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.h:198]: Interesting protocol!

File: src/ray/core_worker/pubsub/publisher.h
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.h:191]:
```suggestion
  ///   - If the subscriber is dead, there will be no new long polling connection coming in
  ///     again. Otherwise, the long polling connection will be reestablished.
```

File: src/ray/core_worker/pubsub/publisher.h
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.h:193]:
```suggestion
  ///   subscriber as dead.
```

**Review by clarkzinzow — COMMENTED** [review]
A few comments on a first pass!

Also, at a higher level, parts of the protocol seem a bit brittle due to their implicit nature, e.g. sending an empty publish in order to trigger another subscriber long-poll instead of an explicit reconnect RPC. I could imagine someone changing the `HandleLongPollingResponse` implementation in the subscriber without being aware of that part of the protocol and easily breaking it. Probably not a big concern at the moment, but it might be worth it to add more implementation comments in both the publisher and subscriber detailing that protocol, that'd also make it easier for other devs to grok it when reading the code.

**Comment by clarkzinzow** [comment]
Also, are you planning on updating the C++ tests before merging?

**Comment by rkooo567** [comment]
@clarkzinzow Yes! I am working on it now.

File: src/ray/core_worker/pubsub/publisher.h
Comment (rkooo567) [reply src/ray/core_worker/pubsub/publisher.h:29]: This will be part of the third PR! 

- First PR: Current stuff
- Second PR: Moving them to ray_common
- Third PR: Use templated implementation and adapt it for the core worker. 

So you are suggesting to make this ID part also a part of template right?

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:574]: Good point! Let me fix that...

File: src/ray/core_worker/pubsub/publisher.h
Comment (rkooo567) [reply src/ray/core_worker/pubsub/publisher.h:193]: For some reason, lint makes it revert, so I will keep it as is.

**Review by fishbone — COMMENTED** [review]
I'm kind of worrying about the locks here. I feel fine about the lock on the node level, but node+id makes me feel there might be a perf issue. Well, it's just my gut feeling.

Is it possible to encapsulate this in the other place, for example, put id into the message and in core worker, we deal with it like what we did for object fetching? We do it in io_context which is single thread and we can avoid lock there.

File: src/ray/core_worker/pubsub/publisher.h
Comment (rkooo567) [reply src/ray/core_worker/pubsub/publisher.h:29]: Yep. That will be in the scope of the third PR!

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:350]: I think it is better passing `periodical_runner_` because if we just pass `io_service`, we will just reimplement what it is doing. 

Re; ownership. What can go wrong by using shared pointer here? (Asking for curiosity)

Comment (clarkzinzow) [reply src/ray/core_worker/core_worker.cc:350]: I think Yi is referring to the ownership of the periodical runner being shared between the core worker and the pubsub subcomponent instead of keeping the core worker as the unique owner and passing a reference. Doing the latter would make more sense to me (and I think is more idiomatic C++) if the core worker is guaranteed to outlive the pubsub subcomponent, which is the case here, right?

When a class's lifetime encapsulates the lifetime of one of its members, and when that member wishes to share a resource with that class, tying that resource's lifetime to the class's lifetime (i.e. making the class the owner) and giving the member a reference to that resource can provide both an ownership model that's easier to reason about and better performance (no shared pointer reference counting).

Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:350]: Discussed with @iycheng offline. I think performance is not a big deal here because it is passed and used only once. The lifetime stuff makes sense, so I will just do that (since it is not that difficult to implement). But note that this pattern is not followed by other parts (we can probably consider refactoring in the future).

Comment (iycheng) [reply src/ray/core_worker/core_worker.cc:350]: I saw a lot of places where we use shared pointer instead of unique pointer which is a bad pattern that makes the code harder to understand (ownership is vague) and sometimes bringing performance issues.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-unique

Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:350]: @iycheng Do you think it makes sense to organize these things somewhere? (like some of cpp best practices we should follow)?

Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:350]: Maybe it is good to have a doc (with the current state of art in our repo). That can also give us what parts need to be re-written easily

Comment (iycheng) [reply src/ray/core_worker/core_worker.cc:350]: Yes, that's a good idea. We can identify some bad patterns in our code and make a doc about them.

File: src/ray/core_worker/pubsub/publisher.h
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.h:144]: Perfect! 😁

File: src/ray/core_worker/pubsub/publisher.h
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.h:150]: Nit: Could mention that this periodical runner is used to periodically check for dead subscribers, although that's pretty obvious when looking at the constructor code. Up to you!

File: src/ray/raylet/pubsub/subscriber.cc
Comment (clarkzinzow) [inline src/ray/raylet/pubsub/subscriber.cc:26]: Nit: In interest of variables being defined in the smallest scope in which they are used, I'd usually vote for `subscriber_address` to be defined within the conditional, since we only need it when making a long-polling connection.

File: src/ray/core_worker/pubsub/publisher.cc
Comment (clarkzinzow) [inline src/ray/core_worker/pubsub/publisher.cc:236]: Nit: Same comment as below, I'd usually prefer that `subscriber_id` is defined (or just using `it.first` directly) within the `if (disconnect)` scope since it's only used when unregistering a subscriber.

**Review by clarkzinzow — APPROVED** [review]
LGTM, only nits that shouldn't block the PR being merged!

**Comment by rkooo567** [comment]
Failed tests seem to be unrelated. 

Windows: `tests:test_actor_resources ` is very flaky in the master
ASAN: gcs_server_failure => shouldn't be related to this PR
Rllib failure => Failing on master
Serve test failure => flaky on master
Mac wheel => timeout frequently in PR builds

**Pull request merged**
